pkg/services/sqlstore/migrations/ualert/silences.go

```diff
@@ -55,6 +56,46 @@ func (m *migration) addSilence(da dashAlert, rule *alertRule) error {
 	return nil
 }
 
+func (m *migration) addNoDataSilence(da dashAlert, rule *alertRule) error {
```

Comment:
If a user later changes the no data state for this alert from `Keep Last State` to something like `Alerting` do we expire the old silence?

Reply:
Good question! Currently, we do not expire the silence when the rule is changed. However, the silence will not match any alert instances. So, it will become kinda useless. I am hesitant whether we should do that because ngalert does not have that relationship between silence and alert rule: users have to create a rule and then a silence for NoData state separately.

Reply:
Won't the silence match the `DatasourceNoData` alert? I might have misunderstood but I thought we wanted notifications for that alert when we change no data state from `Keep Last State` to something else?